### DISCUSSION TITLE
How to clear all cached data or mutate all keys on a scenario like user logging out?

### DISCUSSION THREAD

**Discussion by GorkemSahin** [vercel/swr#1494] [Q&A]
On previous versions of SWR, we had `cache.clear()` or we could iterate over `cache.keys()` and mutate the ones we want. After SWR 1.0 both seems to be gone. I think it was nice having a simple API to do that.

**Comment by shuding** [comment]
Please check this out: https://swr.vercel.app/docs/advanced/cache#access-to-the-cache. You can still have access to the cache object in 1.0.

**Reply by GorkemSahin** [reply]
@shuding  Yes, but the cache object is typed as follows on` swr/dist/types.d.ts`:

    export interface Cache<Data = any> {

        get(key: Key): Data | null | undefined;

        set(key: Key, value: Data): void;

        delete(key: Key): void;

    }

So there are no `clear` or `keys`.

**Reply by shuding** [reply]
Yeah there's currently a bug of type inference, so you have to cast the type to `any`. It's actually a `Map` instance.

**Reply by GorkemSahin** [reply]
> Yeah there's currently a bug of type inference, so you have to cast the type to `any`. It's actually a `Map` instance.

Shall I report it as an issue or has someone else done that already?

**Reply by shuding** [reply]
Feel free to open an issue, thank you! 🙏

**Reply by Ne0xygene** [reply]
Hello,
I have the same problem,
I opened an issue about it.
Issue : https://github.com/vercel/swr/issues/1502

**Comment by marcinn** [comment]
Hi. 

I'm unable to clear cache on login/logout. I'm using 1.1.0-beta.9 and I've lost possibility to clear sensitive data.
I can access `cache` instance, I can call `clear()`, but it seems like useSWRConfig is returning other instances of cache.

I have App.js similar to:
```
export const App = () => {
   const {config} = useSWRConfig()
   useEffect(() => {
     config.cache.clear() // it is called
   }, [appContext.isAuthenticated])

  // ...

    return (                                                                           
         <AppContext.Provider value={appContext}>                                       
             <SWRConfig value={{provider: () => new Map()}}> 
               {children}
             </SWRConfig>
         </AppContext.Provider>
     )
}
```

Instance of `config` called from useEffect is cleared as expected.  But `useSWRConfig()`, which is placed deeper in the components tree (just close to useSWR - to check state of the cache), is returning data same as before clearing.

It would be nice to clear SWR's cache outside of hooks (for example: inside a reducer), same as with version <1.0. Is possible to bring back this possibility?

**Reply by shuding** [reply]
`useSWRConfig` is accessed out of `<SWRConfig>`, so it returns the global instance not the customized one. If you remove that `{ provider: () => new Map() }` it will work.

**Comment by larsvanpoucke** [comment]
I was looking for this aswell, found it here: https://swr.vercel.app/docs/mutation

> You can use the filter function to clear all cache data, which is useful when logging out:
> ```
> const { mutate } = useSWRConfig()
> 
> ...
> 
> const clearCache = () => mutate(
>   () => true,
>   undefined,
>   { revalidate: false }
> )
> 
> // ...clear cache on logout
> clearCache()
> ```

**Comment by szawruk** [comment]
There is now method for that. Please checkout [https://github.com/vercel/swr/issues/1887#issuecomment-1171269211](https://github.com/vercel/swr/issues/1887#issuecomment-1171269211).
